File: LLMgeneratedCode/llama-1.R
# Install necessary libraries if not already installed
#install.packages(c("dplyr", "caret", "ggplot2"))

# Load required libraries
library(dplyr)
library(caret)
library(ggplot2)

# Load the data and metadata from the Gene Expression Omnibus dataset GSE149440
data <- read.csv("https://www.ncbi.nlm.nih.gov/geo/gdservice.fcgi?accession=gse149440&submit=download")
metadata <- read.table("https://www.ncbi.nlm.nih.gov/geo/gdservice.fcgi?accession=gse149440&submit=download", header = FALSE)

# Convert the metadata to a data frame
colnames(metadata) <- c("sample_id", "description", "chip_type", "protocol_version", "run",
                        "series_id", "study_name", "study_description", 
                        "design_type", "design_description", "experiment_type",
                        "experiment_description", "platform", "platform_description",
                        "platform_version", "study_date", "collection_date", "completion_date",
                        "geographic_region", "research_area", "research_concept_descriptions",
                        "sample_type", "sampleprep_method", "sequencing_library", 
                        "sequence technology", "sequence platform", "sequence version",
                        "sequencing_library_layout", "chip_name", "chip_description",
                        "array_type", "array_description", "array_feature_matrix",
                        "array_feature_types", "chip_accession")

# Create a new data frame that contains only the relevant metadata
training_data <- metadata[metadata$design_type == "RNA-seq", ]

# Create a new data frame that contains only the relevant expression data for training and testing
train_expression_data <- data[, sapply(data, function(x) which(names(x) %in% colnames(training_data)))]
test_expression_data <- train_expression_data[train_expression_data$sample_id %in% training_data$sample_id[training_data$design_type == "RNA-seq" & training_data$ch1 == 0], ]

# Create a new data frame that contains only the gestational age values
gestational_age_data <- metadata[metadata$design_description == "GestationalAge:CH1", ]
predict_gestational_age_values <- as.numeric(gestational_age_data$geographic_region)

# Fit a linear regression model to predict gestational age from expression data
model <- lm(predict_gestational_age_values ~ rowSums(train_expression_data))

# Predict the gestational age values for the test set
predictions <- predict(model, newdata = test_expression_data)

# Calculate the root mean squared error (RMSE) of the predictions on the test set
rmse <- sqrt(mean((test_expression_data$predict_gestational_age_values - predictions)^2))
print(paste("Root Mean Squared Error (RMSE):", rmse))

# Plot a scatter plot of predicted vs actual gestational age values for the test set
ggplot(test_expression_data, aes(x = predict_gestational_age_values, y = predict_gestational_age_values)) +
  geom_point() +
  geom_line(aes(y = test_expression_data$predict_gestational_age_values), color = "red") +
  labs(title = "Predicted vs Actual Gestational Age Values for the Test Set",
       subtitle = paste("RMSE:", rmse),
       x = "Predicted Gestational Age Value",
       y = "Actual Gestational Age Value")

File: LLMgeneratedCode/llama-1.py
import requests
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

# Download GSE149440 dataset from GEO
url = 'https://www.ncbi.nlm.nih.gov/geo/query.cgi?acc=GSE149440&submit=Submit+query'
response = requests.get(url)

# Parse the XML response and extract the expression data
data = pd.read_csv('GSE149440-GEODATA.txt', sep='\t')

# Map gene symbols to their IDs
gene_map = {row['symbol']: row['accession'] for index, row in data.iterrows()}
gene_ids = [gene_map[genename] for genenames in data['feature'].values]

# Get the metadata and create a DataFrame with relevant information
metadata_url = 'https://www.ncbi.nlm.nih.gov/geo/query.cgi?acc=GSE149440&submit=Submit+query'
response = requests.get(metadata_url)
metadata = pd.read_csv('metadata.txt', sep='\t')

# Extract the gestational age data and filter out samples not assigned to training set
gestational_age = metadata['gestation']
train_samples = metadata[metadata['train:ch1'] == '1']['accession'].values

# Split data into features (gene expression) and target variable (gestational age)
X = pd.DataFrame(data[['feature', gene_ids]].T, columns=data['feature']).set_index('accession')
y = gestational_age.loc[train_samples].reset_index(drop=True)

# Fit a linear regression model
X_train, X_test, y_train, y_test = train_test_split(X.loc[X.index.isin(train_samples)], 
                                                    y.loc[y.index.isin(train_samples)], 
                                                    test_size=0.2, random_state=42)
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
predictions = model.predict(X_test)

# Calculate root mean squared error (RMSE) of predictions
rmse = np.sqrt(np.mean((y_test - predictions) ** 2))
print('Root Mean Squared Error:', rmse)

# Generate scatter plot for prediction vs actual values
plt.scatter(y_test, predictions)
plt.xlabel('Actual Gestational Age')
plt.ylabel('Predicted Gestational Age')
plt.title('Gestational Age Prediction Scatter Plot')
plt.savefig("./out2/llama-1.py.pdf", format="pdf")

